File: src/utils/market_hours.py
"""Market hours detection utility for US equity markets."""

from datetime import datetime
from zoneinfo import ZoneInfo


ET = ZoneInfo("America/New_York")


def is_market_open(dt: datetime | None = None) -> bool:
    """Check if US market is open (Mon-Fri 9:30 AM - 4:00 PM ET).

    Args:
        dt: Datetime to check. If naive, treats as ET. Defaults to now.

    Returns:
        True if market is open, False otherwise.
    """
    if dt is None:
        dt = datetime.now(ET)
    elif dt.tzinfo is None:
        # Naive datetime: assume ET
        dt = dt.replace(tzinfo=ET)
    else:
        # Aware datetime: convert to ET
        dt = dt.astimezone(ET)

    # Weekends (0=Sunday, 6=Saturday)
    if dt.weekday() >= 5:
        return False

    # Before 9:30 AM or after 4:00 PM
    if dt.hour < 9 or dt.hour >= 16:
        return False

    # Between 9:00-9:30 AM
    if dt.hour == 9 and dt.minute < 30:
        return False

    return True

File: frontend/src/lib/market-hours.ts
/**
 * Market hours detection for US equity markets.
 * Mon-Fri 9:30 AM - 4:00 PM ET.
 */

/**
 * Convert Date to US/Eastern timezone.
 */
function toET(date: Date): Date {
  // Create a new date in ET by using the timezone offset
  const tz = "America/New_York";
  const et = new Date(date.toLocaleString("en-US", { timeZone: tz }));
  return et;
}

/**
 * Check if US market is currently open.
 * Mon-Fri 9:30 AM - 4:00 PM ET.
 *
 * @param date - Date to check (defaults to now)
 * @returns true if market is open, false otherwise
 */
export function isMarketOpen(date: Date = new Date()): boolean {
  const et = toET(date);

  // Check weekend (0=Sunday, 6=Saturday)
  const day = et.getDay();
  if (day === 0 || day === 6) {
    return false;
  }

  const hour = et.getHours();
  const minute = et.getMinutes();

  // Before 9:30 AM or after 4:00 PM
  if (hour < 9 || hour >= 16) {
    return false;
  }

  // Between 9:00-9:30 AM
  if (hour === 9 && minute < 30) {
    return false;
  }

  return true;
}
